nengo_bones/templates.py

```diff
@@ -0,0 +1,121 @@
+        The heading for the section in the config file containing config
+        options specific to the template being rendered.
+    template_file : str
+        Filename for the input template file.
+    """
+
+    __slots__ = ("output_file", "section", "template_file")
+
+    def __init__(self, output_file):
+
+        self.output_file = output_file
+
+        section = output_file.lstrip(".")
+        section = section.replace(".", "_")
+        section = section.replace("/", "_")
+        section = section.lower()
+        self.section = section
+
+        self.template_file = "%s.template" % (output_file,)
+
+    def render(self, **kwargs):
+        """
+        Render this template to a string.
+
+        Parameters
+        ----------
+        kwargs : dict
+            Will be passed on to the ``template.render`` function.
+        """
+        return env.get_template(self.template_file).render(**kwargs)
+
+    def render_to_file(self, output_dir, output_name=None, **kwargs):
+        """
+        Render a template to file.
+
+        .. note:: Rendered shell scripts (files with the ``.sh extension``)
+                  are automatically marked as executable.
+
+        Parameters
+        ----------
+        output_dir : str
+            Directory in which the rendered file should be placed.
+        output_name : str, optional
+            An alternative filename for the rendered file.
+            This overrides the class's internal ``output_file`` attribute.
+        kwargs : dict
+            Will be passed on to the ``render`` function.
+        """
+        if output_name is None:
+            output_name = self.output_file
+        output_path = os.path.join(output_dir, output_name)
+
+        if not os.path.exists(os.path.dirname(output_path)):
+            os.makedirs(os.path.dirname(output_path))
+
+        with open(output_path, "w") as f:
+            f.write(self.render(**kwargs))
+
+        # We mark all `.sh` files as executable
+        if output_name.endswith(".sh"):
+            st = os.stat(output_path)
+            os.chmod(output_path, st.st_mode | stat.S_IEXEC)
+
+
+def _load_env():
+    bones_toplevel = os.path.normpath(
+        os.path.join(os.path.dirname(__file__))
+    )
+    # Load overridden templates first.
+    # Builtins are referenced with templates/*.template
+    override_dirs = []
+    override_dirs.append(".templates")
+    override_dirs.append(bones_toplevel)
+    override_loader = jinja2.FileSystemLoader(override_dirs)
+    # If those fail, use the builtins
+    builtin_loader = jinja2.FileSystemLoader(
+        os.path.join(bones_toplevel, "templates")
+    )
+
+    jinja_env = jinja2.Environment(
+        loader=jinja2.ChoiceLoader([override_loader, builtin_loader]),
+        trim_blocks=True,
+        lstrip_blocks=True,
+        keep_trailing_newline=True,
+    )
+    jinja_env.filters["rstrip"] = lambda s, chars: s.rstrip(chars)
+    return jinja_env
+
+
+env = _load_env()
```

Comment:
It's not really either, it's just for convenience and shorter code. We could replace it all the uses of it with calls to `load_env` instead, that would be fine by me, just seemed like we have fewer worries about people using it in weird multithreaded situations so might as well do it with the fewest characters. But we can replace with calls to `load_env` if you prefer that.

Reply:
Added a fixup to do this, I've just been burned too many times by global state, I'm paranoid :wink:. I used the click context so that the environment only needs to be set up once per `bones-check/generate` call.